druid/src/widget/images.rs

```diff
@@ -0,0 +1,242 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An Image widget.
+//! Please consider using SVG and the SVG wideget as it scales much better.
+
+use std::error::Error;
+use std::marker::PhantomData;
+use std::str::FromStr;
+
+use image;
+
+use crate::{
+    Affine, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx,
+    PaintCtx, Point, Rect, RenderContext, Size, UpdateCtx, Widget,
+};
+
+use crate::piet::{ImageFormat, InterpolationMode};
+
+// These are based on https://api.flutter.dev/flutter/painting/BoxFit-class.html
+#[derive(Clone, Copy, PartialEq)]
+pub enum FillStrat {
+    Contain,
+    Cover,
+    Fill,
+    FitHeight,
+    FitWidth,
+    None,
+    ScaleDown,
+}
+
+impl Default for FillStrat {
+    fn default() -> Self {
+        FillStrat::Fill
+    }
+}
+
+/// Calulate a orign and scale for a image.
+///
+/// This takes some propities of a widget and a fill Stratagy and returns a origin and scale.
+fn get_scale_offset(parent: Size, fit_box: Size, fit_type: FillStrat) -> (Point, Point) {
+    let scalex = parent.width / fit_box.width;
+    let scaley = parent.height / fit_box.height;
+
+    let scale: Point = match fit_type {
+        FillStrat::Contain => {
+            let scale = scalex.min(scaley);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::Cover => {
+            let scale = scalex.max(scaley);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::Fill => Point {
+            x: scalex,
+            y: scaley,
+        },
+        FillStrat::FitHeight => Point {
+            x: scaley,
+            y: scaley,
+        },
+        FillStrat::FitWidth => Point {
+            x: scalex,
+            y: scalex,
+        },
+        FillStrat::ScaleDown => {
+            let scale = scalex.min(scaley).min(1.0);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::None => Point { x: 1.0, y: 1.0 },
+    };
+
+    let origin_x = (parent.width - (fit_box.width * scale.x)) / 2.0;
+    let origin_y = (parent.height - (fit_box.height * scale.y)) / 2.0;
+    let origin = Point::new(origin_x, origin_y);
+
+    (scale, origin)
+}
+
+/// A widget that renders an Image
+pub struct Image<T> {
+    image_data: ImageData,
+    phantom: PhantomData<T>,
+    fill: FillStrat,
+}
+
+impl<T: Data> Image<T> {
+    /// Create an Image-drawing widget from ImageData.
+    ///
+    /// The Image will scale to fit its box constraints.
+    pub fn new(image_data: ImageData) -> Self {
+        Image {
+            image_data,
+            phantom: Default::default(),
+            fill: FillStrat::default(),
+        }
+    }
+
+    pub fn set_fill(&mut self, newfil: FillStrat) {
+        self.fill = newfil;
+    }
+
+    pub fn fill_mode(mut self, mode: FillStrat) -> Self {
```

Comment:
I would put constructor/builder fns ahead of other fns, and I would document both of these.

Reply:
I have attempted this but not really sure of the right wording

Reply:
what dose fns stand for?

Reply:
```
/// A builder-style method to set the widget's [`FillStrat`].
/// 
/// This determines how the image will be fit to the widget's bounds.
///
/// [`FillStrat`]: enum.FillStrat.html
```